Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeAddress.php

```diff
@@ -0,0 +1,107 @@
+ *
+ * By using the content graph for the content repository
+ * one can build a subgraph with the right perspective to find this node:
+ *
+ *      $subgraph = $contentRepository->getContentGraph($nodeAddress->workspaceName)->getSubgraph(
+ *          $nodeAddress->dimensionSpacePoint,
+ *          VisibilityConstraints::withoutRestrictions()
+ *      );
+ *      $node = $subgraph->findNodeById($nodeAddress->aggregateId);
+ *
+ * @api
+ */
+final readonly class NodeAddress implements \JsonSerializable
+{
+    private function __construct(
+        public ContentRepositoryId $contentRepositoryId,
+        public WorkspaceName $workspaceName,
+        public DimensionSpacePoint $dimensionSpacePoint,
+        public NodeAggregateId $aggregateId,
+    ) {
+    }
+
+    public static function create(
+        ContentRepositoryId $contentRepositoryId,
+        WorkspaceName $workspaceName,
+        DimensionSpacePoint $dimensionSpacePoint,
+        NodeAggregateId $aggregateId,
+    ): self {
+        return new self($contentRepositoryId, $workspaceName, $dimensionSpacePoint, $aggregateId);
+    }
+
+    public static function fromNode(Node $node): self
+    {
+        return new self(
+            $node->contentRepositoryId,
+            $node->workspaceName,
+            $node->dimensionSpacePoint,
+            $node->nodeAggregateId
+        );
+    }
+
+    /**
+     * @param array<string, mixed> $array
+     */
+    public static function fromArray(array $array): self
+    {
+        return new self(
+            ContentRepositoryId::fromString($array['contentRepositoryId']),
+            WorkspaceName::fromString($array['workspaceName']),
+            DimensionSpacePoint::fromArray($array['dimensionSpacePoint']),
+            NodeAggregateId::fromString($array['aggregateId'])
+        );
+    }
+
+    public static function fromJsonString(string $jsonString): self
+    {
+        return self::fromArray(\json_decode($jsonString, true, JSON_THROW_ON_ERROR));
+    }
+
+    public function withAggregateId(NodeAggregateId $aggregateId): self
+    {
+        return new self($this->contentRepositoryId, $this->workspaceName, $this->dimensionSpacePoint, $aggregateId);
+    }
+
+    public function equals(self $other): bool
+    {
+        return $this->contentRepositoryId->equals($other->contentRepositoryId)
+            && $this->workspaceName->equals($other->workspaceName)
+            && $this->dimensionSpacePoint->equals($other->dimensionSpacePoint)
+            && $this->aggregateId->equals($other->aggregateId);
+    }
+
+    public function toJson(): string
+    {
+        return json_encode($this, JSON_THROW_ON_ERROR);
+    }
+
+    public function jsonSerialize(): mixed
+    {
+        return [
+            'contentRepositoryId' => $this->contentRepositoryId,
+            'workspaceName' => $this->workspaceName,
+            'dimensionSpacePoint' => $this->dimensionSpacePoint,
+            'aggregateId' => $this->aggregateId
+        ];
```

Comment:
nitpick:
```suggestion
        return get_object_vars($this);
```
is more stable to future changes and we do that in a lot of other places too

Reply:
ha i guess were 50/50 on this, as i have a `fromArray` i prefer it manually added so its clear that any changes HAVE to be also adjusted in the fromArray... but yeah idk

Reply:
We use `get_object_vars()` in 52 places in the core – I like it but I would not protest if you guys want to change this. But please lets not change the style in every PR

Reply:
I'm team get_object_vars